cmd/kubeadm/app/phases/selfhosting/podspec_mutation.go

```diff
@@ -101,6 +103,26 @@ func setMasterTolerationOnPodSpec(podSpec *v1.PodSpec) {
 	podSpec.Tolerations = append(podSpec.Tolerations, kubeadmconstants.MasterToleration)
 }
 
+// setPodIPOnPodSpec sets the environment variable POD_IP using downward API
+func setPodIPOnPodSpec(podSpec *v1.PodSpec) {
+	envVar := v1.EnvVar{
+		Name: "POD_IP",
+		ValueFrom: &v1.EnvVarSource{
+			FieldRef: &v1.ObjectFieldSelector{
+				FieldPath: "status.podIP",
+			},
+		},
+	}
+
+	podSpec.Containers[0].Env = append(podSpec.Containers[0].Env, envVar)
+
+	for i := range podSpec.Containers[0].Command {
+		if strings.Contains(podSpec.Containers[0].Command[i], "advertise-address") {
+			podSpec.Containers[0].Command[i] = "--advertise-address=$(POD_IP)"
```

Comment:
Isn't it possible for the POD_IP to shift?  This doesn't seem possible if you're are dealing with an HA env.

Reply:
The Pod IP == Node IP when `hostNetwork` is true, which it always is in this case.
Does that make sense to you?

Reply:
This was written assuming pod IP == node IP as Lucas mentioned. Though I do see how using pod IP can lead to confusions down the road as it's not obvious to everyone that kube-apiserver would run in hostNetwork. Maybe we should be explicit and use host IP downward API? WDYT?

Reply:
Yeah I would use host ip for clarity.